sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -361,7 +361,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       connectionProperties: Properties): DataFrame = {
     assertNoSpecifiedSchema("jdbc")
     // connectionProperties should override settings in extraOptions.
-    val params = extraOptions.toMap ++ connectionProperties.asScala.toMap
+    val params = extraOptions ++ connectionProperties.asScala
```

Comment:
shall we call `extraOptions.toMap` here? to be case-preserving

`params` is used to create `new JDBCOptions(url, table, params)`, and `JDBCOptions.asProperties` needs to be case-preserving.

Reply:
and to guarantee `connectionProperties` overrides `extraOptions`, we may also need to do `map1.filterKeys... ++ map2`

Reply:
`extraOptions ++ connectionProperties.asScala` will guarantee `connectionProoperties` overrides the `extraOption`. We added `++` operation for that~

Also, `new JDBCOptions(url, table, params)` works like this. The passed `params: CaseInsensitiveMap` guarantees that JDBC_URL and JDBC_TABLE_NAME override `parameters`.
```
  def this(url: String, table: String, parameters: Map[String, String]) = {
    this(CaseInsensitiveMap(parameters ++ Map(
      JDBCOptions.JDBC_URL -> url,
      JDBCOptions.JDBC_TABLE_NAME -> table)))
  }
```

The above code is for your logic.

Reply:
I'll add a test case for this.

Reply:
ah got it, thanks!